operator/pkg/ciliumidentity/controller_test.go

```diff
@@ -197,3 +199,244 @@ func verifyCIDUsageInCES(ctx context.Context, fakeClient *k8sClient.FakeClientse
+			// Run podActions in the sequence defined in the test case.
+			// Verify CID state after each podAction.
+			for actionIdx, podAction := range tc.podActions {
+				pod := podAction.podObj
+				switch podAction.action {
+				case "create":
+					if _, err := fakeClient.Slim().CoreV1().Pods(pod.Namespace).Create(ctx, pod, metav1.CreateOptions{}); err != nil {
+						t.Fatalf("create pod: %v", err)
+					}
+				case "update":
+					if _, err := fakeClient.Slim().CoreV1().Pods(pod.Namespace).Update(ctx, pod, metav1.UpdateOptions{}); err != nil {
+						t.Fatalf("update pod: %v", err)
+					}
+				case "delete":
+					if err := fakeClient.Slim().CoreV1().Pods(pod.Namespace).Delete(ctx, pod.Name, metav1.DeleteOptions{}); err != nil {
+						t.Fatalf("delete pod: %v", err)
+					}
+				}
+
+				if len(tc.cidUpdateList) > 0 && tc.cidUpdateList[actionIdx] != nil {
+					cidList := cidStore.List()
+					if len(cidList) == 0 {
+						t.Fatal("listed CIDs for update = 0, want >= 1")
+					}
+
+					cidUpdateRef := tc.cidUpdateList[actionIdx]
+					var cidToUpdate *capi_v2.CiliumIdentity
+					// Get from CID store the CID that needs to be updated.
+					for _, cid := range cidList {
+						if cmp.Equal(cid.SecurityLabels, cidUpdateRef.origSecLabels) {
+							cidToUpdate = cid
+							break
+						}
+					}
+
+					if cidToUpdate == nil {
+						t.Fatalf("CID for update not found in CID store, labels: %v", cidUpdateRef.updatedCID.SecurityLabels)
+					}
+
+					cidToUpdate.Labels = cidUpdateRef.updatedCID.Labels
+					cidToUpdate.SecurityLabels = cidUpdateRef.updatedCID.SecurityLabels
+					if _, err := fakeClient.CiliumV2().CiliumIdentities().Update(ctx, cidToUpdate, metav1.UpdateOptions{}); err != nil {
+						t.Fatalf("update CID: %v", err)
+					}
+				}
+
+				var lastRecordedErr error
+				err := testutils.WaitUntil(func() bool {
+					// Ensure that we don't use the same CID twice.
+					// And that all CIDs in the store are used, and there are no leaked CIDs.
+					usedCIDs := make(map[string]*capi_v2.CiliumIdentity)
+					cidList := cidStore.List()
+
+					for _, wantLbls := range tc.expectedInCIDStore[actionIdx] {
+						matchFound := false
+						for _, cid := range cidList {
+							if cmp.Equal(cid.SecurityLabels, wantLbls) {
+								if _, used := usedCIDs[cid.Name]; used {
+									continue
+								}
+
+								usedCIDs[cid.Name] = cid
+								matchFound = true
+								break
+							}
+						}
+						if !matchFound {
+							lastRecordedErr = fmt.Errorf("match for labels not found: %v", wantLbls)
+							return false
+						}
+					}
+
+					for _, cid := range cidList {
+						if _, used := usedCIDs[cid.Name]; !used {
+							lastRecordedErr = fmt.Errorf("leaked CID: %v", cid)
+							return false
+						}
+					}
+
+					return true
+				}, WaitUntilTimeout)
+				if err != nil {
+					t.Errorf("expected CID action=%v not found for labels list=%v, error: %v", podAction, tc.expectedInCIDStore[actionIdx], lastRecordedErr)
```

Comment:
I think we should improve the error message: right now, we are printing two times the expected labels list, one from `lastRecordedErr`, one from `tc.expectedInCIDStore[actionIdx]`. Ideally, we want to output the actual state with the desired state.

Reply:
I updated the log message to include more info as you suggested.